File: android/app/src/main/java/cz/qrplatba/domain/Spayd.kt
package cz.qrplatba.domain

/**
 * Build a SPAYD ("QR Platba", Czech Banking Association) string.
 * Format: SPD*1.0*ACC:<IBAN>*AM:<amount.2dp>*CC:CZK*X-VS:<vs>*MSG:<name>
 *
 * SPAYD field values must not contain '*' (the field separator). The MSG value
 * is sanitized: '*' replaced with space, and per spec MSG is limited to 60 chars.
 */
object Spayd {
    fun build(
        iban: String,
        amount: Money,
        vs: String,
        message: String,
        currency: String = "CZK",
    ): String {
        val msg = sanitizeMessage(message)
        return listOf(
            "SPD",
            "1.0",
            "ACC:$iban",
            "AM:${formatAmount2dp(amount)}",
            "CC:$currency",
            "X-VS:$vs",
            "MSG:$msg",
        ).joinToString("*")
    }

    private fun sanitizeMessage(message: String): String =
        message
            .replace("*", " ")
            .replace(Regex("[\\r\\n\\t]"), " ")
            .trim()
            .take(60)
}

File: server/src/domain/spayd.ts
import { formatAmount2dp, type Money } from "./money.js";

/**
 * Build a SPAYD ("QR Platba", Czech Banking Association) string.
 * Format: SPD*1.0*ACC:<IBAN>*AM:<amount.2dp>*CC:CZK*X-VS:<vs>*MSG:<name>
 *
 * SPAYD field values must not contain '*' (the field separator). The MSG value
 * is sanitized: '*' stripped, and per spec MSG is limited to 60 chars.
 */
export interface SpaydInput {
  iban: string;
  amount: Money;
  vs: string;
  message: string;
  currency?: string;
}

export function buildSpayd(input: SpaydInput): string {
  const currency = input.currency ?? "CZK";
  const msg = sanitizeMessage(input.message);
  const parts = [
    "SPD",
    "1.0",
    `ACC:${input.iban}`,
    `AM:${formatAmount2dp(input.amount)}`,
    `CC:${currency}`,
    `X-VS:${input.vs}`,
    `MSG:${msg}`,
  ];
  return parts.join("*");
}

function sanitizeMessage(message: string): string {
  // Remove field separators and control chars; SPAYD MSG max length is 60.
  return message
    .replace(/\*/g, " ")
    .replace(/[\r\n\t]/g, " ")
    .trim()
    .slice(0, 60);
}
